v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/build-system.ts

```diff
@@ -131,10 +177,27 @@ export class SolidityBuildSystemImplementation implements SolidityBuildSystem {
       },
     );
 
-    const isSuccessfulBuild = results.every(
-      (result) => !this.#hasCompilationErrors(result),
+    const uncachedResults = results.filter((result) => !result.cached);
+    const uncachedSuccessfulResults = uncachedResults.filter(
+      (result) => !this.#hasCompilationErrors(result.compilerOutput),
     );
 
+    // NOTE: We're not waiting for the writes and clean to finish because we
+    // will only care about the result of these operations in subsequent runs
+    void Promise.all(
+      uncachedSuccessfulResults.map(async (result) => {
+        return this.#compilerOutputCache.set(
+          await result.compilationJob.getBuildId(),
+          result.compilerOutput,
+        );
+      }),
+    ).then(() => {
+      return this.#compilerOutputCache.clean();
+    });
```

Comment:
[When should we clean up the cache](https://www.notion.so/nomicfoundation/solidity-compilation-cache-17b578cdeaf580e3ba10c0d907fd6a4d?pvs=4#182578cdeaf5807099dae5e86efdc624) was one of the open questions stated in the design doc.

To do so from the `cleanArtifacts` method was suggested. I do like this suggestion as then all the "clean" operations would be grouped together. However, when working on the final implementation, I realised it might be more beneficial to call the cache clean from build after all.

As you can see in the snippet, this allows us to ensure that clean is only ever called after all the writes have finished. That way we can avoid (or at least postpone) implementing synchronisation mechanisms between writes and cleans in the cache component.

Reply:
I think once we update the cache to include partial recompilation we may want to revisit this, but it works at the moment.